### PR TITLE
Kernel: Overhaul shutdown procedure

### PR DESCRIPTION
**Pull request by kleinesfilmroellchen**
cc @supercomputer7 @IdanHo

# Shutdown improvements part two, electric boogaloo?

For a long time, our shutdown procedure has basically been:
- Acquire big process lock.
- Switch framebuffer to Kernel debug console.
- Sync and lock all file systems so that disk caches are flushed and
  files are in a good state.
- Use firmware and architecture-specific functionality to perform
  hardware shutdown.

This naive and simple shutdown procedure has multiple issues:
- No processes are terminated properly, meaning they cannot perform more
  complex cleanup work. If they were in the middle of I/O, for instance,
  only the data that already reached the Kernel is written to disk, and
  data corruption due to unfinished writes can therefore still occur.
- No file systems are unmounted, meaning that any important unmount work
  will never happen. This is important for e.g. Ext2, which has
  facilites for detecting improper unmounts (see superblock's s_state
  variable) and therefore requires a proper unmount to be performed.
  This was also the starting point for this PR, since I wanted to
  introduce basic Ext2 file system checking and unmounting.
- No hardware is properly shut down beyond what the system firmware does
  on its own.
- Shutdown is performed within the write() call that asked the Kernel to
  change its power state. If the shutdown procedure takes longer (i.e.
  when it's done properly), this blocks the process causing the shutdown
  and prevents any potentially-useful interactions between Kernel and
  userland during shutdown.

In essence, current shutdown is a glorified system crash with minimal
file system cleanliness guarantees.

Therefore, this commit is the first step in improving our shutdown
procedure. The new shutdown flow is now as follows:
- From the write() call to the power state SysFS node, a new task is
  started, the Power State Switch Task. Its only purpose is to change
  the operating system's power state. This task takes over shutdown and
  reboot duties, although reboot is not modified in this commit.
- The Power State Switch Task assumes that userland has performed all
  shutdown duties it can perform on its own. In particular, it assumes
  that all kinds of clean process shutdown have been done, and remaining
  processes can be hard-killed without consequence. This is an important
  separation of concerns: While this commit does not modify userland, in
  the future SystemServer will be responsible for performing proper
  shutdown of user processes, including timeouts for stubborn processes
  etc.
- As mentioned above, the task hard-kills remaining user processes.
- The task hard-kills all Kernel processes except itself and the
  Finalizer Task. Since Kernel processes can delay their own shutdown
  indefinitely if they want to, they have plenty opportunity to perform
  proper shutdown if necessary. This may become a problem with
  non-cooperative Kernel tasks, but as seen two commits earlier, for now
  all tasks will cooperate within a few seconds.
- The task waits for the Finalizer Task to clean up all processes.
- The task hard-kills and finalizes the Finalizer Task itself, meaning
  that it now is the only remaining process in the system.
- The task syncs and locks all file systems, and then unmounts them. Due
  to an unknown refcount bug we currently cannot unmount the root file
  system; therefore the task is able to abort the clean unmount if
  necessary.
- The task performs platform-dependent hardware shutdown as before.

This commit has multiple remaining issues (or exposed existing ones)
which will need to be addressed in the future but are out of scope for
now:
- Unmounting the root filesystem is impossible due to remaining
  references to the inodes /home and /home/anon. I investigated this
  very heavily and could not find whoever is holding the last two
  references.
- Userland cannot perform proper cleanup, since the Kernel's power state
  variable is accessed directly by tools instead of a proper userland
  shutdown procedure directed by SystemServer.

The recently introduced Firmware/PowerState procedures are removed
again, since all of the architecture-independent code can live in the
power state switch task. The architecture-specific code is kept,
however.

### Kernel: Add accessor for mount host custody

There's no reason this information needs to be secret.

### Kernel/FileSystem: Pass last mount point guest inode to unmount prepare

This will be important later on when we check file system busyness.

### Kernel/Ext2: Check and set file system state

This is supposed to detect whether a file system was unmounted
cleanly or not.

### Kernel/VirtualFileSystem: Count bind mounts towards normal FS mountcount

This is correct since unmount doesn't treat bind mounts specially. If we
don't do this, unmounting bind mounts will call
prepare_for_last_unmount() on the guest FS much too early, which will
most likely fail due to a busy file system.

### Kernel/VirtualFileSystem: Allow unmounting via inode and mount path

This pair of information uniquely identifies any mount point, and it can
be used in situations where mount point custodies are not available.

### Kernel/Tasks: Allow Kernel processes to be shut down

Since we never check a kernel process's state like a userland process,
it's possible for a kernel process to ignore the fact that someone is
trying to kill it, and continue running. This is not desireable if we
want to properly shutdown all processes, including Kernel ones.
Process.h now provides a simple macro which exits the process if it is
dying.

### Kernel: Allow relaxing cleanup task rules during system shutdown

Once we move to a more proper shutdown procedure, processes other than
the finalizer task must be able to perform cleanup and finalization
duties, not only because the finalizer task itself needs to be cleaned
up by someone. This global variable, mirroring the early boot flags,
allows a future shutdown process to perform cleanup on its own.

Note that while this *could* be considered a weakening in security, the
attack surface is minimal and the results are not dramatic. To exploit
this, an attacker would have to gain a Kernel write primitive to this
global variable (bypassing KASLR among other things) and then gain some
way of calling the relevant functions, all of this only to destroy some
other running process. The same effect can be achieved with LPE which
can often be gained with significantly simpler userspace exploits (e.g.
of setuid binaries).

### CODE REVIEW

File: Kernel/FileSystem/Plan9FS/FileSystem.cpp [hunk 1]
```diff
@@ -328,6 +328,7 @@ void Plan9FS::thread_main()
 {
     dbgln("Plan9FS: Thread running");
     do {
+        EXIT_KERNEL_TASK_IF_NECESSARY()
```
Comment (gmta) [inline Kernel/FileSystem/Plan9FS/FileSystem.cpp:331]: Why is this part of this commit?

File: Kernel/FileSystem/Ext2FS/FileSystem.cpp [hunk 2]
```diff
@@ -65,6 +65,9 @@ ErrorOr<void> Ext2FS::initialize_while_locked()
         return EINVAL;
     }
 
+    if (super_block.s_state == EXT2_ERROR_FS)
+        dmesgln("Ext2FS: Was not unmounted cleanly, file system may be errorneous!");
```
Comment (gmta) [inline Kernel/FileSystem/Ext2FS/FileSystem.cpp:69]: errorneous -> erroneous

File: Kernel/FileSystem/Ext2FS/FileSystem.h [hunk 3]
```diff
@@ -70,7 +70,7 @@ class Ext2FS final : public BlockBasedFileSystem {
     virtual ErrorOr<void> initialize_while_locked() override;
     virtual bool is_initialized_while_locked() override;
 
-    virtual ErrorOr<void> prepare_to_clear_last_mount() override;
+    virtual ErrorOr<void> prepare_to_clear_last_mount(Inode& mount_guest_inode) override;
```
Comment (gmta) [inline Kernel/FileSystem/Ext2FS/FileSystem.h:73]: This should have been part of the previous commit

File: Kernel/Tasks/PowerStateSwitchTask.cpp [hunk 4]
```diff
@@ -4,8 +4,210 @@
...
+ErrorOr<void> PowerStateSwitchTask::perform_shutdown()
+{
+    // We assume that by this point userland has tried as much as possible to shut down everything in an orderly fashion.
+    // Therefore, we force kill remaining processes, including Kernel processes, except the finalizer and ourselves.
+    dbgln("Killing remaining processes...");
+    Optional<Process&> finalizer_process;
+    Process::all_instances().for_each([&](Process& process) {
+        if (process.pid() == g_finalizer->process().pid()) {
+            finalizer_process = process;
+        }
```
Comment (gmta) [inline Kernel/Tasks/PowerStateSwitchTask.cpp:91]: You can leave out these braces for single-line if-blocks

File: Kernel/FileSystem/Plan9FS/FileSystem.cpp (hunk 1 above)
Comment (kleinesfilmroellchen) [reply Kernel/FileSystem/Plan9FS/FileSystem.cpp:331]: It's very possible I messed something up; I peeled apart the changes manually this afternoon.

File: Kernel/FileSystem/VirtualFileSystem.h [hunk 5]
```diff
@@ -57,6 +57,8 @@ class VirtualFileSystem {
     ErrorOr<void> bind_mount(Custody& source, Custody& mount_point, int flags);
     ErrorOr<void> remount(Custody& mount_point, int new_flags);
     ErrorOr<void> unmount(Custody& mount_point);
+    // Make sure to pass a path where the given inode is actually mounted, or things may go wrong.
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Kernel/FileSystem/VirtualFileSystem.h:60]: I'm not sure how we can check this properly though? An inode can be mounted to multiple paths, i.e. custodies.

File: Kernel/FileSystem/Ext2FS/FileSystem.cpp [hunk 6]
```diff
@@ -532,7 +532,7 @@ unsigned Ext2FS::free_inode_count() const
     return super_block().s_free_inodes_count;
 }
 
-ErrorOr<void> Ext2FS::prepare_to_clear_last_mount()
+ErrorOr<void> Ext2FS::prepare_to_clear_last_mount(Inode& mount_guest_inode)
```
Comment (gmta) [inline Kernel/FileSystem/Ext2FS/FileSystem.cpp:544]: This commit fails to build independently:

```
/serenity/Kernel/FileSystem/Ext2FS/FileSystem.cpp: In member function 'virtual AK::ErrorOr<void> Kernel::Ext2FS::prepare_to_clear_last_mount(Kernel::Inode&)':
/serenity/Kernel/FileSystem/Ext2FS/FileSystem.cpp:535:58: error: unused parameter 'mount_guest_inode' [-Werror=unused-parameter]
  535 | ErrorOr<void> Ext2FS::prepare_to_clear_last_mount(Inode& mount_guest_inode)
      |                                                   ~~~~~~~^~~~~~~~~~~~~~~~~
```

**Comment by kleinesfilmroellchen** [comment]
Crash appears to be unrelated.

File: Kernel/Memory/AddressSpace.cpp [hunk 7]
```diff
@@ -14,6 +14,7 @@
 #include <Kernel/Memory/MemoryManager.h>
 #include <Kernel/Security/Random.h>
 #include <Kernel/Tasks/PerformanceManager.h>
+#include <Kernel/Tasks/PowerStateSwitchTask.h>
```
Comment (gmta) [inline Kernel/Memory/AddressSpace.cpp:17]: This file is only added in the next commit, so this can't work here.

File: Kernel/Tasks/Thread.cpp [hunk 8]
```diff
@@ -23,6 +23,7 @@
 #include <Kernel/Memory/ScopedAddressSpaceSwitcher.h>
 #include <Kernel/Sections.h>
 #include <Kernel/Tasks/PerformanceEventBuffer.h>
+#include <Kernel/Tasks/PowerStateSwitchTask.h>
```
Comment (gmta) [inline Kernel/Tasks/Thread.cpp:26]: Same here

File: Kernel/Tasks/PowerStateSwitchTask.h [hunk 9]
```diff
@@ -4,8 +4,39 @@
...
+#include <Kernel/Forward.h>
+
 namespace Kernel {
 
+enum class PowerStateCommand : uintptr_t {
+    Shutdown,
+    Reboot,
+};
+// We will pass the power state command to the task in place of a void* as to avoid the complications of raw allocations.
+static_assert(sizeof(PowerStateCommand) == sizeof(void*));
```
Comment (gmta) [inline Kernel/Tasks/PowerStateSwitchTask.h:19]: Apologies, either this is new or I just missed this last time: what "complications of raw allocations" are there with this enum? `spawn` goes ahead and creates a `KString` and a new process, not sure what this is trying to achieve?

Comment (kleinesfilmroellchen) [reply Kernel/Tasks/PowerStateSwitchTask.h:19]: Avoid an allocation, and ownership responsibilities. This just makes things simpler.

Comment (gmta) [reply Kernel/Tasks/PowerStateSwitchTask.h:19]: Ah the ownership POV makes sense to me - gotcha.

**Review by gmta — APPROVED** [review]

**Pull request merged**